process/common.go

```diff
@@ -526,33 +517,20 @@ func getHeaderFromPool(
 func getHeaderFromPoolWithNonce(
 	nonce uint64,
 	shardId uint32,
-	cacher storage.Cacher,
-	uint64SyncMapCacher dataRetriever.Uint64SyncMapCacher,
+	headersCacher dataRetriever.HeadersPool,
 ) (interface{}, []byte, error) {
 
-	if cacher == nil || cacher.IsInterfaceNil() {
+	if headersCacher == nil || headersCacher.IsInterfaceNil() {
 		return nil, nil, ErrNilCacher
 	}
-	if uint64SyncMapCacher == nil || uint64SyncMapCacher.IsInterfaceNil() {
-		return nil, nil, ErrNilUint64SyncMapCacher
-	}
 
-	syncMap, ok := uint64SyncMapCacher.Get(nonce)
-	if !ok {
-		return nil, nil, ErrMissingHashForHeaderNonce
-	}
-
-	hash, ok := syncMap.Load(shardId)
-	if hash == nil || !ok {
-		return nil, nil, ErrMissingHashForHeaderNonce
-	}
-
-	obj, ok := cacher.Peek(hash)
-	if !ok {
+	headers, hashes, err := headersCacher.GetHeaderByNonceAndShardId(nonce, shardId)
+	if err != nil {
 		return nil, nil, ErrMissingHeader
 	}
 
-	return obj, hash, nil
+	//TODO what should we do when we get from pool more than one header with same nonce and shardId
```

Comment:
For now is ok, as I checked the code where this method si called, but we need to change something in removeHeaderFromPools form baseBootstrap. We need to remove header by hash and no by nonce and shard, to remove only the targeted header, which has been returned by this method, and not all the headers with this nonce. Should be checked also all the removes done in the new implementation to be sure that for each call we want to remove all the headers with the given nonce and shard, and not only one targeted.

Reply:
done.